### DISCUSSION TITLE
Enterprise Development, when will it be a reality ?

### DISCUSSION THREAD

**Discussion by aarcoraci** [dotnet/maui#16037] [General]
Yes, this is yet another complain discussion but I feel myself in a position where I will very soon need to decide not only for me but for my team and what's more important the company.

I have developed using Xamarin for a while now due to my familiarity with .net and specially C# but it was not until a couple years back where I started working on a very serious, very complex application using Xamarin.

Given Xamarin life will end soon I must provide alternatives to the company and the natural choice would be Maui. About a month ago I started porting the existing application which as I mentioned contains many complex part such as many renderers and a lot of native code.

My experience with Maui has been incredibly painful, it is taking me back to the worst years of Xamarin (I was there from the beginning) but this time I cannot afford putting one of the company's most valuable assets on the line given the incredibly buggy status of this technology.

I would like, if possible, to hear back from other fellow engineers or developers. Do you think we are somewhere close to have a minimum level of stability ? I cannot believe I have to wait until (at least) November to have the keyboard "next" button to go to the next entry or to correctly render a listview without the size going all over the place.

Unfortunately, as much as I love c# and .net (on my college year I was a MSP, so yes, I'm very close to Microsoft) I believe we cannot wait until November and a decision will need to be made to see if we switch to flutter or a more stable framework.

**Comment by Kukulkano** [comment]
Thanks for asking. I'm in a similar situation. We have a working xamarin app that will not support Android 14 and future OS changes any more. And now we plan to migrate to MAUI, but I only read about nightmare experiences and horrible bugs in basic widgets like [Labels](https://github.com/dotnet/maui/issues/16018), [CollectionViews](https://github.com/dotnet/maui/issues/15977) and [TitleView](https://github.com/dotnet/maui/issues/16015).

I am scared to make the switch to MAUI. And I have to make a decision in the next few weeks. The only enterprise alternative seems Flutter or ReactNative. But in my company we do not have any experience in Dart and Flutter... This is so painful :-(

**Reply by aarcoraci** [reply]
We face the same problem. Although I have some dart experience, building a mobile app architecture is no joke. It's a hole new paradigm.

I think an excellent a professional course of action from Microsoft would be to extend Xamarin life until Maui is somewhat stable. Maybe we can do some sort of poll or "signature collection" thing...

Right now Maui is not even close to be a viable stack for us.

**Reply by Xyncgas** [reply]
If a framework isn't satisfying your need, use another one without hesitation, so you can use what's best for you we don't owe microsoft anything

**Reply by aarcoraci** [reply]
I need very very very strong points in order to propose a stack change to the development team and client. With a mature functioning app written in Xamarin we were hopping MAUI was going to be a smooth transition.

Changing Frameworks is not trivial. I', seeing a lot of promotion from Uno Platform but is it mature enough ? I asked about native code to peterfoot but I don't think he understood what I need: he mentioned native code within conditional compilation. Is this is a bit alarming, I would expect something more "similar" to Xamarin like Dependency or Partial classes.

Flutter is a complete different paradigm and we need to master dart which is not an easy task: we need to master not only the whole widget and data detachment but the million state management tools and the method channels for native .

The situation Microsoft is putting us rushing MAUI is **embarrassing** honestly. 

Right now it's a huge bet and I'm very worry since my decision will have a huge impact on how the company continues to grow (or not).

**Reply by Xyncgas** [reply]
For me, I consider whether a framework is for me :
1. Can I do the stuffs I want to do, doesn't matter how. > Because I can write abstractions, so complexity isn't really what's stopping me
2. What stops me would be if I am stuck with components provided by MAUI instead of making my own, or if I can attempt to make my own implementation of grid but it still can't do what I want to do (e.g in a rectangle 90% the area of the screen, centered in middle, put the grid inside with 33% 66% 33% layout horizontally and 15% 70% 15% layout vertically, now I have a bunch of items, I should enumerate my items on these cells in grid, so each items appears in the grid sequentially, I could control the flow of element to be horizontally or fill vertically first in this grid)
3. Between frameworks that let me do what I want, what's easier

![image](https://github.com/dotnet/maui/assets/35312807/414fdfab-b16d-4a27-ac2e-e5af7679d6d6)
I ended up using MAUI.Blazor, written my own framework that can do UI on top of it or other frameworks.

**Reply by yevgeny-sotnikov** [reply]
@aarcoraci are you using Xamarin.Native or Xamarin.Forms? If first you can simply change build configuration from Xamarin.Android/iOS to .net-android/ios and don't worry about MAUI

**Reply by aarcoraci** [reply]
Forms :(

**Comment by giuseppenovielli** [comment]
This is my opinion -> https://github.com/xamarin/xamarin-macios/issues/18409#issuecomment-1624929630

Next week there are incoming a new .NET MAUI Service Release -> https://github.com/dotnet/maui/discussions/16029#discussioncomment-6380243

**Reply by aarcoraci** [reply]
I appreciate you sharing this. Thanks

**Reply by MartyIX** [reply]
> Next week there are incoming a new .NET MAUI Service Release

Personally I expect that the release will contain fixes for regressions. My expectation is that it won't be a "brave release". MAUI for .NET 8 is the one where the effort goes imho. I say that mostly so that one can do an eduacated decision what to do next.

**Comment by aarcoraci** [comment]
I think we should start a petition to ask Microsoft to extend Xamarin life for a year or two. After going over the messages and some of the bugs I think we are risking too much betting on Maui. If extending Xamarin to at least support XCode 15 is not possible we will have no other option that adopt flutter.

**Comment by peterfoot** [comment]
I do share your frustration with the situation with Xamarin end of support and MAUI progress. There is another option you may wish to investigate which will allow you to reuse your .NET codebase - Uno Platform. It's well supported, mature and open-source. It's based on WinUI so although, like MAUI, it involves changes to your views it is an established XAML dialect and there are lots of resources already available. I recently guest wrote some blog posts for them on migrating projects from Xamarin Forms to Uno Platform which help explain some of the differences (https://platform.uno/blog/). Since it's all based on modern .NET it's compatible with the latest iOS and Android releases and you can continue to use your familiar dev tools and much of your code which makes it an easier transition than Flutter.

**Reply by aarcoraci** [reply]
@peterfoot I'll take a look but I would appreciate if you can point me in the right direction on this. We do quite a lot of native API code, mostly low level audio and microphone access to get raw data.

Where can I find documentation on how to access the AVAudioRecorder or MediaRecorder (iOS and Android) ?

**Reply by peterfoot** [reply]
@aarcoraci It's the same way as in Xamarin/MAUI if you're using a library with different platform builds. To call native APIs in the shared code you can use conditional compilation e.g. #if ANDROID

**Reply by aarcoraci** [reply]
Maybe I'm not being clear, I'm not using libraries I need to write low level code (AVAudio, Audio queues, etc.). I need to know what tools are available for this. The UI looks great but without low level platform access the tool wouldn't be an option for us.

If you happen to know where the documentation is for achieving this (ie, what Flutter does with Method Channels) I would greatly appreciate it

**Reply by peterfoot** [reply]
@aarcoraci It's the same .NET runtime as used by MAUI, you can use all those platform APIs directly in the same way. For example I have an open source library for Bluetooth which builds a NuGet containing iOS, Android and Windows dlls. You can consume it from a .NET Maui app or an Uno platform app. If you need P/Invoke or binding libraries you can use those too.

**Reply by Xyncgas** [reply]
Sounds like there's lot of great frameworks by people who are using .NET, they should come here more to talk about what they can do when people are struggling at doing things

**Comment by BernhardPollerspoeck** [comment]
Hi aarcoraci,
i am pretty much in the same position as you and did have the same problems.
As employed AND freelancing i feel its a hard time to provide quick and reliable solutions for customers of all trades.
for this reason i just now started and explore Project Uno (as already linked by peterfoot here) , because maui in one side is just not a production ready platform(even if its marketed as such), on the other hand the update policy is just a show stopper.
For now i just did experiments with the default project, but at least the startup process was pretty smooth compared to maui.
i now could test the template project on iphone, ipad, android, wpf and web, wich is already impressive. But all platforms first try started and worked flawless. 
For support i asked questions on theyre discord and even got some first hand team answers on questions wich felt really satisfying.
After a few more tests i have a full project that i want to create in uno leveraging at least mobile and web as platforms. If you (or someone else) is interested on how this journey goes, feel free to ping me somewhere or just drop me a DM.

**Reply by Xyncgas** [reply]
Uno is awefully similar to MAUI, what's the difference why aren't we using uno if it's better

**Reply by Xyncgas** [reply]
>As employed AND freelancing i feel its a hard time to provide quick and reliable solutions for customers of all trades.

Just out of curiosity, do you feel the same if you were developing a solution that's based on MacOS exclusively and you know you are going to use XCode

**Reply by BernhardPollerspoeck** [reply]
I think (but can not prove) it mostly due to exposure. .net and maui both belong to microsoft and are cross promoting them in every possible way. 
currently i am at the stage of testing out all uno layouts and controls that are built in. this may take a while, but so far it looks even better than the matured and stable Xamarin.Forms. i really like on how the controls look styled and the same (i am using fluent style) on each platform.

**Reply by BernhardPollerspoeck** [reply]
since i am no native developer, and even if i can build in xcode and swift, i am just not really quallified for an answer here. but i have a personal oppinion: i dont really like xcode as development environment. its just really finnicky. but disclaimer: this could just be me not knowing it inside out.

**Reply by Xyncgas** [reply]
tbh, writing UI tends to end up with a lot of code... I needed couple lines of codes for this thing and that thing and they add up, I gotta refactor so things are together and when they need to present something they have what they need, abstraction is usually done without foresight...

I am not ready to rewrite application once a while, having rewritten my app couple times, my approach now usually is really get the picture of what I might do in the long run try to find a philosophy that I can put my commitment in which is something Microsoft lacks in their signaling due to the lack of genius people such as myself

Just kidding

**Comment by developer9969** [comment]
HI, 
same experience as you unfortunately ,  no one can fault the team - the very small team and managers with extremely low resources do an oustanding job considering  but putting this aside is simply a "shitshow!

It was doomed from day one and many people that worked at microsoft and left said that.

 We have xamarin forms apps and therefore the natural progression is Maui. We ported 01 big application and we had no end issues , its plagued with workarounds, I am not going to list them here but what a nightmare it's been and still is. If they fix some issues eventually it might all crumble again... 

What annoys me most is how this all caos has been managed! 

Lets take their current stand:

.Net 7  (very little is being ported as team is worried and rightly so about breaking it even more) 
.Net 8 All fixes are there..  but can you run it against a device using windows host and hot restart  - NO
Can you release or deploy it NO 

We get some weird stuff when deployed in .net 7 , pages that work in debug on physical iPHone and then deployed pages do not load! Whats going on!!!!

Well, Uno - avalonia etc..  not sure will go down well with my managers if I say we need to do another rewrite.. 

We just hope that they fix the bloody basics.. 

The reality I cant just pack up and move to flutter or any other platform with being given time to do so..  

In short 

Maui  is 100%  NOT  READY FOR PRODUCTION UNLESS YOU ARE READY TO PUT A MILLION PLASTERS!!

**Reply by Xyncgas** [reply]
>It was doomed from day one and many people that worked at microsoft and left said that.

Who

**Reply by Xyncgas** [reply]
If you are affirmative you can't do your job in MAUI why not switch to another framework, talk about this with your boss say you can't do it in MAUI

In china, if you mention you can't do something, you are blamed instead of figuring out a solution based on reality.

Iet's hope the situation is presented correctly as it needs to be to the decision maker, if they say you can't program and taunt at you by suggesting to place another guy to do your job, it would be a sign it's like china

**Comment by samirgcofficial** [comment]
Tested a basic collection view in .net 8 collection view selection item is broken 😞. And that’s the future of dotnet maui from .7 to .8

**Reply by Xyncgas** [reply]
👏

**Reply by aarcoraci** [reply]
my feeling is.net 8 won't be good enough, there are so many bugs for so many basic controls. I honestly don't care they take all the time they need, just don't terminate Xamarin support !

**Comment by Xyncgas** [comment]
If Xamarin is not supported, it might be possible for you to stick to the last version forever right, just don't update to MAUI don't port it just keep using Xamarin forever for that application. Eventually you can rewrite it, when there's forces pushing to

**Reply by mohachouch** [reply]
The problem is that from November 2024, it will be impossible to update an application because Xamarin Forms only supports Android 13 while the Android store will require Android 14 as a target sdk

**Reply by Xyncgas** [reply]
Can you ask people to not use Android 14, if the application's that important for people they wouldn't update their android, but on the other hand you can be risking losing the people who are your customers and doesn't care about relying on your application enough that they want to update to android 14 no matter what and complain to you about it not working.

Although, dropping support doesn't mean it won't work actually, if you are lucky your app can still run on the latest version of android

**Reply by mohachouch** [reply]
I think you misunderstood the problem. even if we want to stay on android 13. We will no longer be able to update this version on the android store from November 2024. So no more patches or new features. It is simply the death of the application. So we have no choice: either we migrate to Maui but it is not stable or we have to change technology.

**Reply by aarcoraci** [reply]
it's quite the dilema. For enterprise solutions asking people not to use something or just drop a framework is not easy. There is money involve, teams to coordinate and new technologies to master.

I will keep insisting on them taking all the time they need to work on MAUI but while they do it keep Xamarin with enough support.

In my case I have elevated the issues and we are considering flutter, and if it is up to me, I honestly rather take these months building the core of the app in flutter than trying to work on the current version of MAUI

**Reply by mohachouch** [reply]
But here we come back to the problem that Microsoft does not want to invest in maintaining Xamarin. The team is not sized to manage two frameworks. they are already having trouble stabilizing Maui.

**Reply by mohachouch** [reply]
The only solution is for the community to fork Xamarin Forms and release the next version of Android and iOS. I studied the impacts. The problem being the support of the new net7 and net 8 framework. If we decide to migrate Xamarin forms on net 7, it will also be necessary to migrate all the third-party libraries. the work is enormous. Second solution, is to fork xamarin android, from latest stable version is only add new version but it is also complex.

**Comment by developer9969** [comment]
ok guys, do you know what is the most frustrating thing of them all "SILENCE" you would expect some words of reassurance from MS  but here we are speculating and being frustated... 

Do you only way MS will extend  xamarin life is if a big client moans big time- otherwise is just workarounds in Maui and hope for the best.

I just hope that next preview is good enough to move away from .net 7,

**Comment by yevgeny-sotnikov** [comment]
From my point of view we still can to develop using Xamarin.Native which will be simply rebranded to .net-android and .net-ios without any dramatical chаnges inside. But yes, with Xamarin.Forms and MAUI we have much worse situation
